src/_guides/language/effective-dart/design.md

```diff
@@ -880,26 +880,47 @@ var buffer = new StringBuffer()
 
 ## Type annotations
 
-In Dart 1.x, adding static types to your variables is optional.
+A "type annotation" is a place in your code where you write down the static type
+of a variable, field, function return type, parameter, etc. Type annotations are
+how you tell the type checker your constraints on the kinds of values that flow
+into different parts of your code.
 
-{% include optional-types-2.0.html %}
+In most places, Dart allows you to omit a type annotation and infers a type for
+you based on the nearby context (or defaults to the `dynamic` type). Inference
+is a powerful tool to spare you the effort of writing (and reading) types that
+are obvious or uninteresting, but it's still worth explicitly annotating in some
+cases. The guidelines here strike the best balance we've found between brevity
+and explicitness, flexibility and safety.
 
-{% comment %}
-update-for-dart-2.0
-{% endcomment %}
+The fact that Dart has both type inference and an actual `dynamic` type leads to
+some confusion about what it means to say code is "untyped". Does that mean the
+code is dynamically typed, or that you didn't *write* the type? To avoid that
+confusion, we avoid saying "untyped" and instead use:
 
-### DO type annotate public APIs.
+*   If the code is "type annotated", the type was explicitly written in the
+    code.
 
-Type annotations are important documentation for how a library should be used.
-Annotating the parameter and return types of public methods and functions helps
-users understand what the API expects and what it provides.
+*   If the code is "inferred", no type annotation was written down, and Dart
+    successfully figured out the type on its own. Inference can fail, in which
+    case the guidelines don't consider that "inferred". In some places,
+    inference failure is a static error. In others, Dart uses `dynamic` as the
+    fallback type.
+
+*   If the code is "dynamic", then its static type is the special `dynamic`
+    type. Code can be explicitly annotated `dynamic` or it can be inferred.
+
+*   If the code has any other type, it's not "dynamic". Code can end up with a
+    non-`dynamic` type by being explicitly annotated with some other type, or by
+    having Dart infer a non-`dynamic` type.
+
```

Comment:
In all the cases above, rather than "If the code...", how about "If the declaration..." (because, I believe, type annotations appear in declarations, not in arbitrary code).

Reply:
How about "expression" rather than "code"?

Reply:
Actually, nm, "code" is fine.